### PR TITLE
Add `//impeller/typographer`.

### PR DESCRIPTION
**Pull request by chinmaygarde**
Renders shaped text frames. Has the ability to plug into to different text
shapers and render glyphs using different techniques.

For now, the Aiks layer expects a prepared glyph atlas. But this will be changed
so that the render pass will be responsible for preparing these and setting them on
the content renderer after pass consolidation.

### CODE REVIEW

File: display_list/display_list_dispatcher.h [hunk 1]
```diff
@@ -116,6 +116,9 @@ class DisplayListDispatcher final : public flutter::Dispatcher {
                                 SkScalar mwz,
                                 SkScalar mwt) override;
 
+  // |flutter::Dispatcher|
+  void transformReset() override;
```
Comment (bdero) [inline display_list/display_list_dispatcher.h:120]: transformReset doesn't appear to exist in the dispatcher on engine head, so I got a build failure due to the override here. Is there a corresponding engine PR this is intended to build against?

Comment (chinmaygarde) [reply display_list/display_list_dispatcher.h:120]: Ugh, yeah. I'll back it out.

File: display_list/display_list_dispatcher.cc [hunk 2]
```diff
@@ -6,6 +6,9 @@
 
 #include "flutter/fml/trace_event.h"
 #include "impeller/geometry/path_builder.h"
+#include "impeller/typographer/backends/skia/text_frame_skia.h"
```
Comment (dnfield) [inline display_list/display_list_dispatcher.cc:9]: this include seems unused, and the other changes in this file seem unrelated to typography/text...

File: display_list/display_list_dispatcher.h (hunk 1 above)
Comment (chinmaygarde) [reply display_list/display_list_dispatcher.h]: Done.

File: entity/contents.cc [hunk 3]
```diff
@@ -457,4 +457,114 @@ bool ClipRestoreContents::Render(const ContentContext& renderer,
...
+  VS::BindFrameInfo(cmd, pass.GetTransientsBuffer().EmplaceUniform(frame_info));
+
+  // Common fragment uniforms for all glyphs.
+  FS::BindGlyphAtlasSampler(
+      cmd,                                                        // command
+      atlas_->GetTexture(),                                       // texture
+      renderer.GetContext()->GetSamplerLibrary()->GetSampler({})  // sampler
+  );
+
+  // Common vertex information for all glyphs.
```
Comment (dnfield) [inline entity/contents.cc:510]: It'd be helpful to me if this comment explained why we're addding a dummy 0x1 rect here

**Comment by bdero** [comment]
The Aiks examples have a little bit of funk going on for me. Is that expected at the moment?
image

**Comment by chinmaygarde** [comment]
Yeah, xMin, yMin, xMax & yMax aren't being accounted for the in the metrics. I added to the test to track the fix I am working on. I can back it out if you wish.

**Review by dnfield — APPROVED** [review]
Overall LGTM

**Comment by bdero** [comment]
Nah just checking to make sure it's known and not just me. We should totally keep it in. :)

File: entity/contents.cc (hunk 3 above)
Comment (chinmaygarde) [reply entity/contents.cc:510]: The rect is not a dummy. Since we are drawing glyphs individually (to be batched later), we don't want to give each glyph unique vertex information. So we give all glyphs the same vertices but each gets its size in its uniform data. The vertex shader then transforms this unit sized box correctly. This also helps in the fragment shader where the interpolated vertex information (already unit-sized) is used to sample the texture from the atlas.

Comment (dnfield) [reply entity/contents.cc:510]: Ahhh I see, we're basically giving it a reference point with this?

**Review by bdero — APPROVED** [review]

File: entity/contents.cc (hunk 3 above)
Comment (chinmaygarde) [reply entity/contents.cc:510]: I added a clarifying comment. The unit-sized box is doing triple duty. 1: Making it so that all glyphs get the same vertex data (and hence the data needs to be generated just once), 2: Actually specifying a quad so that the glyph can be drawn, and, 3: Helping out the fragment shader by not making it need additional coordinates to sample from the glyph atlas and making it easier (without having to apply an inverse scale) to just apply a scale-perspective transform for the glyph with an offset.

**Comment by chinmaygarde** [comment]
Yeah, the font metrics right now only have an ascent and descent. I'm fixing that and the color and styling stuff now.
